common/meta/grpc/src/grpc_client.rs

```diff
@@ -491,25 +474,124 @@ impl MetaGrpcClient {
+                    return Ok(c);
+                }
+                Err(e) => {
+                    tracing::warn!(
+                        "grpc_client create channel with {} faild, err: {:?}",
+                        addr,
+                        e
+                    );
+                    label_counter_with_val_and_labels(
+                        META_GRPC_MAKE_CLIENT_FAILED,
+                        vec![(LABEL_ENDPOINT, addr.to_string())],
+                        1,
+                    );
+                    {
+                        let mut ue = self.unhealthy_endpoints.lock().unwrap();
+                        ue.insert(addr.to_string(), ());
+                    }
+                    if i == eps.len() - 1 {
+                        // reach to last addr
+                        return Err(e);
+                    }
+                    continue;
+                }
+            }
+        }
+        Err(MetaError::InvalidConfig("endpoints is empty".to_string()))
+    }
+
     pub fn endpoints_non_empty(endpoints: &[String]) -> std::result::Result<(), MetaError> {
         if endpoints.is_empty() {
             return Err(MetaError::InvalidConfig("endpoints is empty".to_string()));
         }
         Ok(())
     }
 
+    async fn get_endpoints(&self) -> Vec<String> {
+        let eps = self.endpoints.read().await;
+        (*eps).clone()
+    }
+
     #[tracing::instrument(level = "debug", skip(self))]
     pub async fn set_endpoints(
         &self,
         endpoints: Vec<String>,
     ) -> std::result::Result<(), MetaError> {
         Self::endpoints_non_empty(&endpoints)?;
 
+        let distinct_cnt = endpoints.iter().filter(|n| !(*n).is_empty()).count();
+        if distinct_cnt < endpoints.len() / 2 + 1 {
+            tracing::warn!(
+                "distinct endpoints small than majority of meta cluster nodes {}<{}, endpoints: {:?}",
+                distinct_cnt,
+                endpoints.len(),
+                endpoints
+            );
+            return Ok(());
+        }
+
         let mut eps = self.endpoints.write().await;
         *eps = endpoints;
         Ok(())
     }
 
+    #[tracing::instrument(level = "debug", skip(self))]
+    pub async fn sync_endpoints(&self) -> std::result::Result<(), MetaError> {
+        let channel = self.make_channel().await?;
+
+        let mut client = MetaServiceClient::new(channel.clone());
+        let endpoints = client
+            .member_list(Request::new(MemberListRequest {
+                data: "".to_string(),
+            }))
+            .await?;
+        let result: Vec<String> = endpoints.into_inner().data;
+        self.set_endpoints(result).await?;
+        Ok(())
+    }
+
+    async fn auto_sync_endpoints(self: Arc<Self>, mut cancel_tx: OneSend<()>) {
+        loop {
+            select! {
+                _ = cancel_tx.closed() => {
+                    return;
+                }
+                _ = sleep(Duration::from_secs(5)) => {
```

Comment:
`endpoints` changes rarely, hence the sleep time could be relatively longer than several seconds IMHO.